### DISCUSSION TITLE
Welcome to Atomico Discussions!

### DISCUSSION THREAD

**Discussion by UpperCod** [atomicojs/atomico#45] [General]
## 👋 Welcome Atomico!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by ghost** [comment]
Hi,

I'm trying to make atomicojs to prerender the web component.
But its seems to render the web component content twice.
What would be the easiest way to make sure atomicojs doesn't render twice does adding id's help?

Ex:
https://codesandbox.io/s/sparkling-waterfall-fql9v?file=/index.html:0-746

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta http-equiv="X-UA-Compatible" content="ie=edge" />
    <title>Static Template</title>
  </head>
  <body>
    <my-component name="123">
      <div>
        Name: 123
      </div>
    </my-component>
    <script type="module">
      import { c, html } from "https://unpkg.com/atomico"; // 4.0kB
      function component({ name }) {
        return html`<host>
          <div>
            Name: ${name}
          </div>
        </host>`;
      }
      component.props = {
        name: String
      };
      customElements.define("my-component", c(component));
    </script>
  </body>
</html>
```

Btw atomicojs is awesome was previously using haunted + lit-html for my project.
Also is it possible to not use virtual dom and just do direct dom replacement like this library https://github.com/forgojs/forgo?

**Reply by UpperCod** [reply]
Hi, thanks for your question, apparently you are trying to rehydrate the DOM (Load the existing content as part of the DOM for Atomico), the problem is that Atomico works the DOM as fragments at a higher level, example

```html
<my-component>
  <h1>DOM outside Atomico</h1>
  <!--fragment: start-->
  <h1>DOM inside Atomico</h1>
  <!--fragment: end-->
</my-component>
```

to avoid conflict with the work of the DOM by other libraries, for example a Liberia X (Angular, React or Lit) creates fragment marks, if Atomico will apply DOM reuse it would eliminate or replace the fragment marks of the X library in the process of comparing nodes breaking the coexistence. This also improves the interaction with polyfills of webcomponents like CSSStyleSheet, since Atomico would not parse nodes outside of your fragment.

I'm attentive to all the support you need for your project!

**Reply by ghost** [reply]
Thanks got it now that you use fragments and host to implement the vdom diffing. Anyways I'm using atomico.js in a SSG to render the web component content statically. I did get the ssr part working. But the client doesn't rehyrdate.
https://github.com/pyros2097/density-ssg

I did try to work from here in renderChildren,
https://github.com/pyros2097/density-ssg/blob/master/bin/templates/atomico.js#L444

and prevent insertBefore in the parent and instead check if the dom node already exists and just do the diffProps instead and also create the fragments. But couldn't figure it out. Will try another take at it.

Thanks for your quick reply.

**Reply by UpperCod** [reply]
Happy to work with you on an SSR implementation and there are ways to achieve it.

but for this I must add inheritance of the symbolId, let me explain ...

Atomico's render differs from other libraries by the host tag. Libraries do not normally work with containers, but Atomico uses the same render cycle to manipulate the container and this forces Atomico to create an ID for each render
In order to avoid having conflicts due to rewriting of events or fragments, the id of the render is exposed in the webcomponent in the symbolId property https://github.com/atomicojs/atomico/blob/master/src/element/custom-element.js#L49

in theory we should replace that line with this:

```js
this.symbolId =   this.symbolId  || Symbol();
```

this property creates or retrieves an object that represents the last state of the DOM
https://github.com/atomicojs/atomico/blob/master/src/render.js#L164

to associate the SSR you have to create the symbolId of the component and create the fragments manually and associate them with the webcomponent., example:

```js
export function ssr(component) {
  const symbolId = Symbol();
  component.symbolId = symbolId;
  const s = new Mark("");
  const e = new Mark("");
  component.prepend(s);
  component.append(e);
  component[symbolId] = { fragment: { s, e } };
}

```

This is just the start with Atomico, it could be rehydrated and even the state of the hooks

**Reply by ghost** [reply]
Got it this was the node[id] in the render. Seeing that ssr actually requires a dom with prepend and append in nodejs. I'm not completely sure how to serialize it so that it rehydrates. I will have a go with this library and see if I can serialise it properly. https://github.com/WebReflection/linkedom
Thanks

**Reply by UpperCod** [reply]
Sorry for the delay in my answer, I have already added support for symbolId inheritance.

**Reply by ghost** [reply]
Thanks UpperCod. For now I have gone with a slightly modified version of lit-html even though it doesn't have rehydration as it renders on top of the existing dom tree and that solved my problem for now. It also doesn't use a vdom which was one of my main interests.

**Reply by UpperCod** [reply]
I understand that you use a different tool, but I just finished [supporting SSR and DOM hydration for Atomico](https://github.com/atomicojs/atomico/releases/tag/1.25.0) I invite you to try this feature to see if it is useful to you

**Example:** For this you just have to add the attribute `data-hydrate` to the component that you want to hydrate.
  
```html
<my-component name="123" data-hydrate><div>Name: 123</div></my-component>
```
https://codesandbox.io/s/atomico-hydrate-lwqtv

**Reply by UpperCod** [reply]
I apologize for having announced it in advance SSR, I am still working on the support of this feature, I need to work on a functional example to share that utility, regarding hydration this api is stable, so it will not have variations.

**Reply by ghost** [reply]
Thanks. I will look into it and see if it helps.
